_source/_posts/2021-06-18-native-java-comparison.adoc

```diff
@@ -0,0 +1,683 @@
+----
+
+=== Make a Micronaut App from Scratch
+
+You might be wondering, "how did you build a secure Micronaut app"? Did I just hide the complexity? No, it only takes five steps to create the same app.
+
+1. Use https://sdkman.io[SDKMAN!] to install Micronaut's CLI:
+
+   sdk install micronaut
+
+2. Create an app using the `mn create-app` command and rename the project's directory:
+
+   mn create-app com.okta.rest.app --build maven
+   mv app micronaut
+
+3. Add Micronaut's libraries for JWT security:
++
+[source,xml]
+----
+<dependency>
+    <groupId>io.micronaut.security</groupId>
+    <artifactId>micronaut-security</artifactId>
+</dependency>
+<dependency>
+    <groupId>io.micronaut.security</groupId>
+    <artifactId>micronaut-security-jwt</artifactId>
+</dependency>
+----
+
+4. Add a `HelloController` in `src/main/java/com/okta/rest/controller`:
++
+[source,java]
+----
+package com.okta.rest.controller;
+
+import io.micronaut.http.MediaType;
+import io.micronaut.http.annotation.Controller;
+import io.micronaut.http.annotation.Get;
+import io.micronaut.http.annotation.Produces;
+import io.micronaut.security.annotation.Secured;
+import io.micronaut.security.rules.SecurityRule;
+
+import java.security.Principal;
+
+@Controller("/hello")
+public class HelloController {
+
+    @Get
+    @Secured(SecurityRule.IS_AUTHENTICATED)
+    @Produces(MediaType.TEXT_PLAIN)
+    public String hello(Principal principal) {
+        return "Hello, " + principal.getName() + "!";
+    }
+
+}
+----
+
+5. Enable and configure JWT security in `src/main/resources/application.yml`:
++
+[source,yaml]
+----
+micronaut:
+  ...
+  security:
+    enabled: true
+    token:
+      jwt:
+        enabled: true
+        signatures:
+          jwks:
+            okta:
+              url: https://{yourOktaDomain}/oauth2/default/v1/keys
+----
+
+That's it! Now you can start the app or build the native image as shown above.
+
+Now let's take a look at Quarkus.
+
+== Run a Quarkus Java API
+
+Open a terminal, cd into the `quarkus` directory, and run `quarkus:dev` to start the app.
+
+[source,shell]
+----
+cd quarkus
+./mvn quarkus:dev
+----
+
+Update the URLs in `src/main/resources/application.properties` to use your Okta domain.
+
+[source,properties]
+----
+mp.jwt.verify.publickey.location=https://{yourOktaDomain}/oauth2/default/v1/keys
+mp.jwt.verify.issuer=https://{yourOktaDomain}/oauth2/default
+----
+
+=== Test Your Quarkus API with HTTPie
+
+Generate or copy an access token from OpenID Connect <debugger/> and use it to test your Quarkus API.
+
+[source,shell]
+----
+http :8080/hello Authorization:"Bearer $TOKEN"
+----
+
+You should see your email in the response.
+
+image::{% asset_path 'blog/native-java-comparison/httpie-quarkus-bearer-token.png' %}[alt=HTTPie call to Quarkus's /hello with bearer token,width=800,align=center]
+
+Did you notice that Quarkus hot-reloaded your `application.properties` file updates? Pretty slick, eh?!
+
+=== Build a Native Quarkus App
```

Comment:
Hello @sberyozkin! I tried switching from `quarkus-smallrye-jwt` to `quarkus-oidc`, adjusting `HelloResource` to use the ID token, and adding `quarkus.oidc.auth-server-url` to `application.properties`. When I try to hit it with an access token, it fails with:

```
2dd9-1: io.quarkus.oidc.OIDCException: Tenant configuration has not been resolved
        at io.quarkus.oidc.runtime.OidcAuthenticationMechanism.resolve(OidcAuthenticationMechanism.java:61)
        at io.quarkus.oidc.runtime.OidcAuthenticationMechanism.authenticate(OidcAuthenticationMechanism.java:40)
```

Is this because a client ID is required? If so, I'm hesitant to use it since none of the other frameworks require a client ID. The only thing that needs a client ID in this post is the SPA app I created for oidcdebugger.com.

Here's a PR to show what I changed: https://github.com/oktadev/native-java-examples/pull/2

Reply:
Hi @mraible Thanks for giving it a try :-). Handling ID tokens and hence the code flow directly requires a [different configuration](https://quarkus.io/guides/security-openid-connect-web-authentication). In this mode it would not handle the bearer tokens unless the application type is `hybrid` in which bdifferent configurationoth bearer and code flow processes can be handled at the same time.
`client-id` is no longer required in the 2.0.0.x line. We encourage having it since there are cases where the users may want to introspect the bearer tokens or even acquire a user info with the bearer tokens. But it is optional otherwise for the bearer tokens but is required for the code flow processes.

I'm not sure why it fails, it can be traced in the logs (the auto-discovery might've failed - I'll check on my own against the address you used) but perhaps it is indeed not worth it at this moment of time - can I propose that you follow up with describing how various frameworks handle the authorization code flow ? I will help with getting a Quarkus application ready.

Cheers

Reply:
Thanks for the detailed analysis @sberyozkin! I'm going to leave things as they are for now. I'm happy to update the post once Quarkus 2.0.0 is released.

Can you confirm that the Quarkus issuer/JWKS look-up happens on the first request rather than at startup?

Reply:
@mraible, np, sounds good

> Can you confirm that the Quarkus issuer/JWKS look-up happens on the first request rather than at startup?

Yes for `quarkus-smallrye-jwt`. 
`quarkus-oidc` does it at startup by default but can be made to fetch the keys on the first request via a dynamic tenant resolution feature.

Thanks